src/rulesets-new/operation-rules.ts

```diff
@@ -0,0 +1,414 @@
+      },
+    );
+  },
+});
+
+const parameterCase = new OperationRule({
+  name: "operation parameters snake case",
+  docsLink: links.standards.parameterNamesPathComponents,
+  matches: (operation, ruleContext) =>
+    !isBreakingChangeAllowed(ruleContext.custom.changeVersion.stability),
+  rule: (operationAssertions) => {
+    operationAssertions.pathParameter.added(
+      "use the correct case",
+      (pathParameter) => {
+        const name = pathParameter.value.name;
+        if (snakeCase(name) !== name) {
+          throw new RuleError({
+            message: `expected parameter name ${name} to be snake case`,
+          });
+        }
+      },
+    );
+
+    operationAssertions.queryParameter.added(
+      "use the correct case",
+      (queryParameter) => {
+        const name = queryParameter.value.name;
+        if (snakeCase(name) !== name) {
+          throw new RuleError({
+            message: `expected parameter name ${name} to be snake case`,
+          });
+        }
+      },
+    );
+  },
+});
+
+const noPutHttpMethod = new OperationRule({
+  name: "no put method",
+  rule: (operationAssertions) => [
+    operationAssertions.added("not use put method", (operation) => {
+      if (operation.method === "put") {
+        throw new RuleError({
+          message: "put is not allowed in JSON:API",
+        });
+      }
+    }),
+  ],
+});
+
+const preventOperationRemoval = new OperationRule({
+  name: "prevent operation removal",
+  docsLink: links.versioning.breakingChanges,
+  matches: (operation, ruleContext) =>
+    !isBreakingChangeAllowed(ruleContext.custom.changeVersion.stability),
+  rule: (operationAssertions) => {
+    operationAssertions.removed("not be allowed", () => {
+      throw new RuleError({
+        message: "expected operation to be present",
+      });
+    });
+  },
+});
+
+const requireVersionParameter = new OperationRule({
+  name: "require version parameter",
+  docsLink: links.versioning.versionParameter,
+  matches: (operation, ruleContext) =>
+    !isBreakingChangeAllowed(ruleContext.custom.changeVersion.stability),
+  rule: (operationAssertions) => {
+    operationAssertions.requirement.hasQueryParameterMatching({
+      name: "version",
+    });
+  },
+});
+
+const tenantFormatting = new OperationRule({
+  name: "tenant formatting",
+  docsLink: links.standards.orgAndGroupTenantResources,
+  matches: (operation, ruleContext) =>
+    !isBreakingChangeAllowed(ruleContext.custom.changeVersion.stability),
+  rule: (operationAssertions) => {
+    operationAssertions.pathParameter.requirement(
+      "use UUID for org_id or group_id",
+      (parameter) => {
+        const name = parameter.value.name;
+        if (name === "group_id" || name === "org_id") {
+          if (!parameter.value.schema) {
+            throw new RuleError({
+              message: "expected parameter to have a schema",
+            });
+          }
+
+          if (
+            !("$ref" in parameter.value.schema) &&
```

Comment:
Please remind me, are refs guaranteed to be resolved? We may easily reuse org and group parameters with $refs.

Reply:
Yes, with a caveat with circular references (circular references will have a `$ref` value)

In the circular reference case, as long as we dereference at _least_ one level, rules will run against the dereferenced values.

Right now, optic-ci doesn't let you use circular references, in the next version release, we'll allow circular references with at least one level dereferenced, so you will always run rules on at least one level deep.

Do you know if there's any case you use (or will use) circular refs. I'm not sure if all the rules here have been written to handle these cases.

Reply:
I think we've avoided circular references... that sounds wild. No, we haven't tried that. Not sure we would want to!

Now I'm curious, are there valid and useful applications for circular refs in OpenAPI? I don't believe I've ever seen this in the wild.

Reply:
I think there are usecases for graph APIs. It's fairly uncommon to see it in regular REST APIs, but it does happen and we don't want optic-ci to crash when we detect one circular ref!